### DISCUSSION TITLE
Upload files expo/react-native

### DISCUSSION THREAD

**Discussion by brunowilliang** [pocketbase/pocketbase#1916] [Q&A] [answered]
I'm trying to create/update my collection's logo field, it creates it but doesn't send the logo, can someone help me.
what is the correct format to send the file? base64 or uri?

my code is this:

const handlePickLogo = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 1,
    });

    const formData = new FormData();

    formData.append('logo', {
      uri: result?.assets[0].uri,
      // type: 'image/jpeg' || 'image/png',
      name: 'logo.jpg',
    });

    if (data) {
      await api.collection(Collections.CompanyProfile).update(data.id, {
        logo: formData,
      });
    } else {
      await api.collection(Collections.CompanyProfile).create({
        "idCompany": company.id,
        "logo": formData,
      })
    }

    if (!result.canceled) {
      return setLogo(result.assets[0].uri);
    }

  };

**Comment by ganigeorgiev** [comment]
In your code you are mixing both json and `FormData`.

File upload is support only via `multipart/form-data`, meaning that you have to send `FormData` object as body and that the file field must be `Blob` (or any of its subclass like `File`).

For a plain JS example please check https://pocketbase.io/docs/files-handling/#uploading-files.

**Comment by brunowilliang** [comment]
@ganigeorgiev 
I tried that way and without success too... :/

`formData.append('logo', {
      uri: new Blob([result?.assets[0].uri]),
      type: 'image/jpeg' || 'image/png',
      name: 'logo.jpg',
    });

    if (data) {
      await api.collection(Collections.CompanyProfile).update(data.id, {
        logo: formData,
      });
    } else {
      await api.collection(Collections.CompanyProfile).create({
        "idCompany": company.id,
        formData,
      })
    }`

**Reply by ganigeorgiev** [reply]
I'm not sure that I understand the above code.

You can't generate a Blob by just passing a url. You need the file content (aka. bytes).

And again, you are still mixing json with FormData, aka. you need to pass only the FormData object as body, not an object with it.

```js
formData = new FormData();

formData.append("idCompany", company.id);
formData.append("yourFileField", file);

await pb.collection("example").create(formData);
```

**Comment by brunowilliang** [comment]
I went the way you told me...
follow the code:
```js
  const handlePickLogo = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.All,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 1,
    });

    const formData = new FormData();

    formData.append("idCompany", company.id);
    formData.append("logo", logo);

    if (data) {
      await api.collection(Collections.CompanyProfile).update(data.id, {
        formData,
      });
    } else {
      await api.collection(Collections.CompanyProfile).create(formData)
      .catch((error) => {
        console.json(error);
      });
    }

    if (!result.canceled) {
      return setLogo(result.assets[0].uri);
    }

  };

```

and got this error in the console @ganigeorgiev 
```json
 {
   "url": "https://reserves.fly.dev/api/collections/companyProfile/records",
   "status": 400,
   "data": {
      "code": 400,
      "message": "Failed to load the submitted data due to invalid formatting.",
      "data": {
         "logo": {
            "code": "validation_unknown_filenames",
            "message": "The field contains unknown filenames."
         }
      }
   },
   "isAbort": false,
   "originalError": null,
   "name": "ClientResponseError 400"
} 
```

**Reply by ganigeorgiev** [reply]
I really don't know how to explain it better.

What is `logo` in the above? **In needs to be `File` or `Blob` instance.**
I'm not familiar with ReactNative and the Expo components and I'm not sure how you can generate it.

Additionally, for the update call you still continue to mix json with FormData. The correct call should be `update(data.id, formData)`

**Comment by brunowilliang** [comment, marked as answer]
@ganigeorgiev 
my function is sending the image:

```js
  const handlePickLogo = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 1,
    });

    if (!result.canceled) {
      const fileUri = result.assets[0].uri;
  
      const formData = new FormData();
      formData.append("idCompany", company.id);
      formData.append("logo", new File([fileUri], "logo.png", { type: "image/png" || "image/jpg" || "image/jpeg" }));
  
      if (data) {
        await api.collection(Collections.CompanyProfile).update(data.id, formData);
      } else {
        await api.collection(Collections.CompanyProfile).create(formData)
          .then(() => {
            Toast({titulo: "Sucesso",descricao: "Logo atualizada com sucesso!",type: "success"});
          })
          .catch((error) => {
            console.json(error);
          });
      }
  
      return setLogo(result.assets[0].uri);
    }

  };
```

but in my backend it appears like this :/
![Screenshot 2023-02-25 at 13 59 43](https://user-images.githubusercontent.com/41432581/221369745-16ce8371-9a95-48cb-805e-159cdd91b177.png)

**Reply by ganigeorgiev** [reply]
It seems that the constructed `File` is not correct.

From the documentation of [`File`](https://developer.mozilla.org/en-US/docs/Web/API/File/File) constructor, the first argument must be:

> An [iterable](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols#the_iterable_protocol) object such as an [Array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array), having [ArrayBuffer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer)s, [TypedArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/TypedArray)s, [DataView](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/DataView)s, [Blob](https://developer.mozilla.org/en-US/docs/Web/API/Blob)s, strings, or a mix of any of such elements, that will be put inside the [File](https://developer.mozilla.org/en-US/docs/Web/API/File). Note that strings here are encoded as UTF-8, unlike the usual JavaScript UTF-16 strings.

If `fileUri` is just a local url/path string then this will not work. You need the file content, aka. bytes. Try to google how to get a blob/bytes content from the result of Expo's `launchImageLibraryAsync`.

**Reply by ganigeorgiev** [reply]
Also the expression `{ type: "image/png" || "image/jpg" || "image/jpeg" }` doesn't make sense, since it will always result to the object `{ type: "image/png" }`

**Reply by brunowilliang** [reply]
@ganigeorgiev 

I'm trying this way but it keeps giving error, I don't know what else to do :/
```js
  const handlePickLogo = async () => {
    let result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [4, 4],
      quality: 1,
    });

    if (!result.canceled) {
      const fileUri = result.assets[0].uri;
      const fileExtension = fileUri.split('.').pop();
      if (!['png', 'jpg', 'jpeg'].includes(fileExtension!.toLowerCase())) {
        return Toast({titulo: "Erro",descricao: "Formato de imagem inválido!",type: "danger"});
      }

      const response = await fetch(fileUri);
      const blob = await response.blob();
  
      const formData = new FormData();
      formData.append("idCompany", company.id);
      formData.append("logo", new File([blob], `logo.${fileExtension}`));
  
      if (data) {
        await api.collection(Collections.CompanyProfile).update(data.id, formData);
      } else {
        await api.collection(Collections.CompanyProfile).create(formData)
          .then(() => {
            Toast({titulo: "Sucesso",descricao: "Logo atualizada com sucesso!",type: "success"});
          })
          .catch((error) => {
            console.json(error);
          });
      }
  
      return setLogo(result.assets[0].uri);
    }

  };
```

**Comment by brunowilliang** [comment]
@ganigeorgiev 

I'm getting this mistake now, can you help me?

```js
 {
   "message": "Request failed with status code 400",
   "name": "AxiosError",
   "stack": "AxiosError@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:699675:29\nsettle@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:700767:37\nonloadend@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:700647:29\ndispatchEvent@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:30065:31\nsetReadyState@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:28706:29\n__didCompleteResponse@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:28499:29\nemit@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:2310:40\n__callFunction@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:21321:36\n@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:21071:31\n__guard@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:21269:15\ncallFunctionReturnFlushedQueue@http://192.168.0.12:19000/node_modules/expo/AppEntry.bundle?platform=ios&dev=true&hot=false:21070:21\ncallFunctionReturnFlushedQueue@[native code]",
   "config": {
      "transitional": {
         "silentJSONParsing": true,
         "forcedJSONParsing": true,
         "clarifyTimeoutError": false
      },
      "adapter": [
         "xhr",
         "http"
      ],
      "transformRequest": [
         null
      ],
      "transformResponse": [
         null
      ],
      "timeout": 0,
      "xsrfCookieName": "XSRF-TOKEN",
      "xsrfHeaderName": "X-XSRF-TOKEN",
      "maxContentLength": -1,
      "maxBodyLength": -1,
      "env": {},
      "headers": {
         "Accept": "application/json, text/plain, */*",
         "Content-Type": "multipart/form-data"
      },
      "method": "post",
      "url": "https://reserves.fly.dev/api/collections/companyProfile/records",
      "data": {
         "_parts": [
            [
               "idCompany",
               "0k4hqsjg6x7rujg"
            ],
            [
               "logo",
               {
                  "uri": "file:///Users/brunowilliang/Library/Developer/CoreSimulator/Devices/6AB65A80-E9EE-4D4D-9CAC-4A980C593A6E/data/Containers/Data/Application/CA8A0DDF-F97E-4D3E-AF44-2C801660D0FA/Library/Caches/ExponentExperienceData/%2540brunowilliang%252Freserves/ImagePicker/63FAEB53-1314-419C-9894-4D825D31D751.jpg",
                  "name": "logo.jpg",
                  "type": "image/jpg"
               }
            ]
         ]
      }
   },
   "code": "ERR_BAD_REQUEST",
   "status": 400
}
```

**Reply by ganigeorgiev** [reply]
@Brunowilliang I'm confused. Are you using `axios` for sending requests or the SDK?

Could you provide a running minimal reproducible repo?

**Reply by brunowilliang** [reply]
@ganigeorgiev 
I tried using the SDK and the axios, I can't provide the code because it's under contract :/

**Comment by brunowilliang** [comment]
it worked, I was uploading an image larger than 5mb which I set as a limit in the backend, sorry for bothering you @ganigeorgiev

**Reply by emaahh** [reply]
hi can you show a working code, i have the same error but my image is under 5mb

**Reply by clodi95** [reply]
> hi can you show a working code, i have the same error but my image is under 5mb

You need to follow this: https://github.com/pocketbase/pocketbase/discussions/2002#discussioncomment-5254168

**Comment by MDReal32** [comment]
Additionally u can use expo's [FileSystem](https://docs.expo.dev/versions/latest/sdk/filesystem/) package/api.
